sarracen/tests/test_interpolate.py

```diff
@@ -62,16 +62,38 @@ def test_interpolate3dcross():
     # first, test a cross-section at z=0
     image = interpolate3DCross(sdf, 'x', 'y', 'z', 'P', CubicSplineKernel(), 0, 0.1, 0.1, -2, -2, 40, 40)
 
-    # should be exactly the same as for a 2D rendering
+    # should be exactly the same as for a 2D rendering, except q values are now taken from the 3D kernel.
     assert image[0][0] == 0
-    assert image[20][0] == approx(CubicSplineKernel().w(np.sqrt((-1.95) ** 2 + 0.05 ** 2), 2), rel=1e-8)
-    assert image[20][20] == approx(CubicSplineKernel().w(np.sqrt(0.05 ** 2 + 0.05 ** 2), 2), rel=1e-8)
-    assert image[12][17] == approx(CubicSplineKernel().w(np.sqrt(0.75 ** 2 + 0.25 ** 2), 2), rel=1e-8)
+    assert image[20][0] == approx(CubicSplineKernel().w(np.sqrt((-1.95) ** 2 + 0.05 ** 2), 3), rel=1e-8)
+    assert image[20][20] == approx(CubicSplineKernel().w(np.sqrt(0.05 ** 2 + 0.05 ** 2), 3), rel=1e-8)
+    assert image[12][17] == approx(CubicSplineKernel().w(np.sqrt(0.75 ** 2 + 0.25 ** 2), 3), rel=1e-8)
 
     # next, test a cross-section at z=0.5
     image = interpolate3DCross(sdf, 'x', 'y', 'z', 'P', CubicSplineKernel(), 0.5, 0.1, 0.1, -2, -2, 40, 40)
 
     assert image[0][0] == 0
-    assert image[20][0] == approx(CubicSplineKernel().w(np.sqrt((-1.95) ** 2 + 0.05 ** 2 + (0.5 ** 2)), 2), rel=1e-8)
-    assert image[20][20] == approx(CubicSplineKernel().w(np.sqrt(2 * (0.05 ** 2) + (0.5 ** 2)), 2), rel=1e-8)
-    assert image[12][17] == approx(CubicSplineKernel().w(np.sqrt(0.75 ** 2 + 0.25 ** 2 + (0.5 ** 2)), 2), rel=1e-8)
+    assert image[20][0] == approx(CubicSplineKernel().w(np.sqrt((-1.95) ** 2 + 0.05 ** 2 + (0.5 ** 2)), 3), rel=1e-8)
+    assert image[20][20] == approx(CubicSplineKernel().w(np.sqrt(2 * (0.05 ** 2) + (0.5 ** 2)), 3), rel=1e-8)
+    assert image[12][17] == approx(CubicSplineKernel().w(np.sqrt(0.75 ** 2 + 0.25 ** 2 + (0.5 ** 2)), 3), rel=1e-8)
+
+
+def test_interpolate3d():
+    df = pd.DataFrame({'x': [0.25],
+                       'y': [0.25],
+                       'z': [0],
+                       'P': [0.5],
+                       'h': [0.125],
+                       'rho': [0.5],
+                       'm': [0.01],
+                       'A': [3]})
+    sdf = SarracenDataFrame(df, params=dict())
+
+    image = interpolate3D(sdf, 'x', 'y', 'A', CubicSplineKernel(), 0.05, 0.05, 0, 0, 10, 10, 10000)
+
+    # w = 0.01 / (0.5 * 0.125^3) = 10.24
+
+    assert image[0][0] == 0
+    # 10.24 * 0.125 * 3 * F(sqrt(0.025^2 + 0.225^2)/0.125) ~= 3.84 * 0.000409322579272
+    assert image[0][4] == approx(0.0015718842, rel=1e-4)
+    # 10.24 * 0.125 * 3 * F(sqrt(0.025^2 + 0.025^2)/0.125) ~= 3.84 * 0.427916515256
+    assert image[5][5] == approx(1.6431994186, rel=1e-4)
```

Comment:
Try to calculate as much of this in the code as you can, rather than doing it by hand and hard coding the value. You can calculate w from the dataframe easily enough, for example.